test/ThecosomataETH.ts

```diff
@@ -0,0 +1,215 @@
+      wethForTreasury,
+      {
+        value: wethForTreasury,
+      }
+    );
+    await wrapForTreasuryTx.wait();
+    const wrapForHelperTx = await curveHelper.wrapAndTransfer(curveHelper.address, wethForHelper, {
+      value: wethForHelper,
+    });
+    await wrapForHelperTx.wait();
+
+    // Populate mock reserve
+    const auditReserveTx = await redactedTreasury.auditReserves();
+    await auditReserveTx.wait();
+
+    // Init pool
+    const initPoolTx = await curveHelper.initPool(wethForHelper, btrflyForHelper);
+    await initPoolTx.wait();
+  });
+
+  describe('checkUpkeep', () => {
+    it('Should not request upkeep if BTRFLY balance is 0', async () => {
+      const btrflyBalance = await btrfly.balanceOf(thecosomata.address);
+      const upkeepNeeded = await thecosomata.checkUpkeep();
+
+      expect(btrflyBalance).to.equal(0);
+      expect(upkeepNeeded).to.equal(false);
+    });
+
+    it('Should request upkeep if BTRFLY balance is above 0', async () => {
+      const mintBtrflyTx = await btrfly.mint(thecosomata.address, btrflyForThecosomata);
+      await mintBtrflyTx.wait();
+
+      const upkeepNeeded = await thecosomata.checkUpkeep();
+
+      expect(upkeepNeeded).to.equal(true);
+    });
+  });
+
+  describe('performUpkeep', () => {
+    it("Should add liquidity using the treasury's WETH and available BTRFLY", async () => {
+      const treasuryPoolTokenBalanceBeforeUpkeep = await curveHelper.poolTokenBalance(
+        redactedTreasury.address
+      );
+      await thecosomata.performUpkeep();
+      const treasuryPoolTokenBalanceAfterUpkeep = await curveHelper.poolTokenBalance(
+        redactedTreasury.address
+      );
+
+      expect(treasuryPoolTokenBalanceAfterUpkeep).to.be.gt(treasuryPoolTokenBalanceBeforeUpkeep);
+    });
+
+    it('Should add liquidity up to the ETH cap in treasury and burn the excess BTRFLY', async () => {
+      // Mint 2x more BTRFLY than remaining available ETH in treasury
+      const ethAmount = ethers.utils.parseUnits('1', 18);
+      const mintBtrflyTx = await btrfly.mint(thecosomata.address, btrflyForThecosomata);
+      await mintBtrflyTx.wait();
+
+      // Calculate the equivalent amount of BTRFLY based on the remaining ETH
+      // Price oracle can change based on timestamp, always fetch latest price to test
+      const poolPrice = await curveHelper.poolPrice();
+      const btrflyAmount = ethAmount.mul(ethers.utils.parseUnits('1', 18)).div(poolPrice)
+        .div(ethers.utils.parseUnits('1', 9));
+      const unusedBtrfly = btrflyForThecosomata.sub(btrflyAmount);
+
+      await expect(
+        thecosomata.performUpkeep()
+      )
+        .to.emit(thecosomata, 'AddLiquidity')
+        .withArgs(
+          ethAmount,
+          btrflyAmount,
+          unusedBtrfly
+        );
+    });
+
+    it("Should not perform upkeep on insufficient balance on either token", async () => {
+      // Mint a very small amount of BTRFLY, which would result in 0 amount in ETH
```

Comment:
This test seems to work as intended (based on the test description) but not because the BTRFLY amount is so small that there is a 0 amount of ETH; `performUpkeep` reverts because there is no ETH left in the treasury so `ethLiquidity` is 0. If the test is run in isolation it actually fails.

Wanted to point this out to confirm that it's in line with what you're expecting.

Reply:
Good catch, thanks! Will update the unit test